### DISCUSSION TITLE
Rabit-MQ setup and It role in multi node setup

### DISCUSSION THREAD

**Discussion by shreedhar-ww** [elsa-workflows/elsa-core#2576] [General]
we have done the below setup for ELSA multinode in the dev enviornment. 

Can you explain the role of Rabit MQ in this setup with a few sample examples? 

Any link/documentation will be helpful

![Multinode Setup](https://user-images.githubusercontent.com/10514279/145517925-a6e4cadb-b09b-49e0-a908-bfaf230de7ce.png)

In this setup, we are creating workflow instances through HTTP call from the API application.

I just wanted to know instead of HTTP can we use some other mechanism?

**Comment by sfmskywalker** [comment]
The role of RabbitMQ is to send messages between ELSA nodes in the cluster. For example, when a workflow definition is published, ALL nodes need to rebuild their trigger indices.

Another example is the cache signaling. Elsa uses local memory caches to store certain information in-memory. But when this cache needs to be invalidated, it needs to happen on all nodes in the cluster. The message broker is used to send this instruction to all nodes.

> In this setup, we are creating workflow instances through HTTP call from the API application.
>
> I just wanted to know instead of HTTP can we use some other mechanism?

In principle, you can use any transport you like to create workflow instances. But, out of the box, Elsa comes only with HTTP endpoints for external applications to interface with. But you're free to e.g. implement service bus message workers that receive messages from your API application, or gRPC, GraphQL, or anything else that you prefer.

You would implement these listeners in the ELSA application, and internally invoke the Elsa library services to create workflow instances.
A great example of this is of course the [Execute API endpoint](https://github.com/elsa-workflows/elsa-core/blob/master/src/server/Elsa.Server.Api/Endpoints/Workflows/Execute.cs) exposed by your ELSA application.

As you will see, it uses the `IWorkflowLaunchpad` service to invoke workflows.
